File: ios/SecretaryVaultAccess/Sources/SecretaryVaultAccessTesting/ManualFlushScheduler.swift
import Foundation
import SecretaryVaultAccess

/// In-memory `FlushScheduler`. Tests inspect `scheduledDelay` and call `fire`.
public final class ManualFlushScheduler: FlushScheduler {
    public private(set) var scheduledDelay: Duration?
    public private(set) var cancelCount = 0
    private var pending: (@MainActor (MonotonicInstant) -> Void)?

    public init() {}

    public func schedule(after delay: Duration,
                         _ work: @escaping @MainActor (MonotonicInstant) -> Void) {
        scheduledDelay = delay
        pending = work
    }

    public func cancel() {
        cancelCount += 1
        scheduledDelay = nil
        pending = nil
    }

    /// Test hook: fire the pending work with a chosen instant (no-op if cancelled).
    @MainActor public func fire(at instant: MonotonicInstant) {
        let work = pending
        pending = nil
        scheduledDelay = nil
        work?(instant)
    }
}

File: android/vault-access/src/test/kotlin/org/secretary/sync/ManualFlushScheduler.kt
package org.secretary.sync

import kotlin.time.Duration

/**
 * In-memory [FlushScheduler] for host tests: the test controls when the pending work
 * fires and with which instant via [fire]. Models a one-shot timer — [fire] clears the
 * pending work before invoking it, so work that re-schedules (the monitor's re-arm) sets
 * a fresh pending entry.
 */
class ManualFlushScheduler : FlushScheduler {
    var scheduledDelay: Duration? = null
        private set
    var cancelCount: Int = 0
        private set
    private var pending: ((MonotonicInstant) -> Unit)? = null

    val hasPending: Boolean get() = pending != null

    override fun schedule(after: Duration, work: (MonotonicInstant) -> Unit) {
        scheduledDelay = after
        pending = work
    }

    override fun cancel() {
        pending = null
        cancelCount++
    }

    /** Fire the pending work at [at]. */
    fun fire(at: MonotonicInstant) {
        val work = pending ?: error("fire called with nothing scheduled")
        pending = null
        work(at)
    }
}
